sw/src/service-worker/service-worker.js

```diff
@@ -0,0 +1,447 @@
+    setDefaultHandler(new NetworkAndTryCache())
+
+    // * 2. Service Worker event listeners
+
+    self.addEventListener('message', event => {
+        if (!event.data) return
+
+        // This allows the web app to trigger skipWaiting via
+        // registration.waiting.postMessage({type: 'SKIP_WAITING'})
+        // Paired with `clientsClaim()` at top of file.
+        if (event.data.type === swMsgs.skipWaiting) {
+            self.skipWaiting()
+        }
+
+        if (event.data.type === swMsgs.startRecording) {
+            startRecording(event)
+        }
+
+        if (event.data.type === swMsgs.completeRecording) {
+            completeRecording(event.source.id) // same as FetchEvent.clientId
+        }
+    })
+
+    // Open DB on activation
+    self.addEventListener('activate', event => {
+        event.waitUntil(createDB().then(removeUnusedCaches))
+    })
+
+    // * 3. Helper functions:
+
+    function urlMeetsDefaultCachingCriteria(url) {
+        // Don't cache if pwa.caching.omitExternalRequests in d2.config is true
+        if (OMIT_EXTERNAL_REQUESTS && url.origin !== self.location.origin)
+            return false
+
+        // Don't cache if url matches filter in pattern list from d2.config.js
+        const urlMatchesFilter = URL_FILTER_PATTERNS.some(pattern =>
+            new RegExp(pattern).test(url.pathname)
+        )
+        if (urlMatchesFilter) return false
+
+        return true
+    }
+
+    /** Called upon SW activation */
+    function createDB() {
+        dbPromise = openSectionsDB()
+        return dbPromise
+    }
+
+    /** Called upon SW activation */
+    async function removeUnusedCaches() {
+        const cacheKeys = await caches.keys()
+        return Promise.all(
+            cacheKeys.map(async key => {
+                const isWorkboxKey = /workbox/.test(key)
+                const isInKeepList = !!CACHE_KEEP_LIST.find(
+                    keepKey => keepKey === key
+                )
+                const db = await dbPromise
+                const isASavedSection = !!(await db.get(SECTIONS_STORE, key))
+                if (!isWorkboxKey && !isInKeepList && !isASavedSection) {
+                    console.debug(
+                        `[SW] Cache with key ${key} is unused and will be deleted`
+                    )
+                    return caches.delete(key)
+                }
+            })
+        )
+    }
+
+    // Triggered on 'START_RECORDING' message
+    function startRecording(event) {
+        console.debug('[SW] Starting recording')
+        if (!event.data.payload?.sectionId)
+            throw new Error('[SW] No section ID specified to record')
+
+        const clientId = event.source.id // clientId from MessageEvent
+        // Throw error if another recording is in process
+        if (isClientRecording(clientId))
+            throw new Error(
+                "[SW] Can't start a new recording; a recording is already in process"
+            )
```

Comment:
Should this queue another recording instead?  Maybe not, probably fine as it is!

Reply:
I think for the purpose of this MVP it would be good to stick to one-recording-at-a-time since that's what the dashboard calls for, but implementing a queue in the future would be cool!